linked_list/linked_list.py

```diff
@@ -9,67 +9,148 @@ def __init__(self, value, next_node = None):
+        self.head = None # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
+        else:
+            return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_first(self, value):
-        pass
+        if self.head == None:
+            new_node = Node(value=value, next_node=None)
+            self.head= new_node
+        else:
+            new_node = Node(value=value, next_node=self.head)
+            self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
     def search(self, value):
-        pass
+        if self.head == None:
+            return False
+        else:
+            current = self.head
+            while current != None:
+                if current.value == value:
+                    return True
+                else:
+                    current = current.next
+            return False
+                
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        if self.head == None:
+            return 0
+        else:
+            counter = 0
+            current = self.head
+            while current != None:
+                counter += 1
+                current = current.next
+            return counter 
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        if self.head == None:
+            return None
+        else:
+            counter = 0 
+            current = self.head
+            while current != None:
+                if counter == index:
+                    return current.value
+                else:
+                    counter += 1
+                    current = current.next
+            return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        else:
+            current = self.head
+            while current != None:
+                if current.next == None:
+                    return current.value
+                else:
+                    current = current.next

```

Comment:
Can be simplified a bit.

```suggestion
            while current.next != None:
                current = current.next
            return current.value
```